### PR TITLE
Update to latest versions of dependencies and scala-2.12.x

### PR DESCRIPTION
**Pull request by wemrysi**

### CODE REVIEW

**Comment by wemrysi** [comment]
Breaking as the latest `shims` version seems to have affected implicit resolution in some situations.

File: impl/src/test/scala/quasar/impl/logging/ResponseBodyLoggingMiddlewareSpec.scala [hunk 1]
```diff
@@ -53,7 +53,8 @@ object ResponseBodyLoggingMiddlewareSpec extends Specification
   "log zero chunks" >> {
     Ref.of[IO, String]("") flatMap { ref =>
       for {
-        _ <- client(0, ref).fetch(GET(Uri.uri("")))(_.as[String])
+        req <- GET(Uri.uri(""))
```
Comment (dcsobral) [inline impl/src/test/scala/quasar/impl/logging/ResponseBodyLoggingMiddlewareSpec.scala:56]: The deprecation warning recommends `uri""`.

File: frontend/src/main/scala/quasar/frontend/logicalplan/Optimizer.scala [hunk 2]
```diff
@@ -104,6 +102,10 @@ final class Optimizer[T: Equal]
   import set._
   import structural._
 
+  // Import shims.monadToScalaz results in ambiguous implicits for other types
+  private implicit def catsStateTScalazMonad[A] = shims.monadToScalaz[State[A, ?]]
+  private implicit def catsEvalScalazMonad[A] = shims.monadToScalaz[Eval]
```
Comment (dcsobral) [inline frontend/src/main/scala/quasar/frontend/logicalplan/Optimizer.scala:107]: I'm slightly concerned about the maintenance overload of all these private conversion definitions. Could this be fixed by unimporting something from a wildcard?

File: project/Versions.scala [hunk 3]
```diff
@@ -1,33 +1,31 @@
...
-  val algebraVersion      = "2.0.0"
-  val argonautVersion     = "6.3.0-M2"
-  val catsVersion         = "2.1.0"
-  val catsEffectVersion   = "2.1.0"
-  val disciplineVersion   = "0.11.1"
+  val algebraVersion      = "2.0.1"
+  val argonautVersion     = "6.3.0-M2" // The latest version that still depends on scalaz-7.2.x
+  val catsVersion         = "2.2.0"
+  val catsEffectVersion   = "2.2.0"
+  val disciplineVersion   = "1.1.0"
```
Comment (dcsobral) [inline project/Versions.scala:8]: On a personal note, I wish we would get rid of this file. I'm not entirely opposed to it, but I dislike very much having two ways of encoding versions across projects, particular on projects that are very active.

**Review by dcsobral — CHANGES_REQUESTED** [review]
I'm not sure `Uri.uri("")` is deprecated -- I used a method like that at one point and it was, but there's a thousand ways to create an URI, and a dozen different URIs. If it is deprecated, I think it would be best to fix both deprecations at the same time.

I'll leave it as request changes. If you say that one isn't deprecated, consider it approved and merge it.

File: project/Versions.scala (hunk 3 above)
Comment (wemrysi) [reply project/Versions.scala:8]: Yeah, I'm not opposed to getting rid of it in favor of constants in `build.sbt`.

File: impl/src/test/scala/quasar/impl/logging/ResponseBodyLoggingMiddlewareSpec.scala (hunk 1 above)
Comment (wemrysi) [reply impl/src/test/scala/quasar/impl/logging/ResponseBodyLoggingMiddlewareSpec.scala]: Oh, I completely overlooked that, we must have deprecation warnings disabled in tests as they aren't emitted. Anyhow, fixed.

**Review by dcsobral — APPROVED** [review]

File: impl/src/test/scala/quasar/impl/logging/ResponseBodyLoggingMiddlewareSpec.scala (hunk 1 above)
Comment (dcsobral) [reply impl/src/test/scala/quasar/impl/logging/ResponseBodyLoggingMiddlewareSpec.scala]: There's a good reason to not have deprecation warnings on tests: if a library deprecates something, it still want to keep testing the deprecated code until it is removed.

File: frontend/src/main/scala/quasar/frontend/logicalplan/Optimizer.scala (hunk 2 above)
Comment (wemrysi) [reply frontend/src/main/scala/quasar/frontend/logicalplan/Optimizer.scala:107]: Yeah, it isn't ideal, and I'm honestly not sure what exactly is causing the ambiguity (there are no `cats` implicit imports in this file) so this seemed the most explicit way of avoiding it.

Importing `shims._` does work, however it also quadruples the compile time which doesn't seem worth it to me and is why we try and avoid `import shims._` wherever we can.

**Pull request merged**